### PR TITLE
[14.0][IMP] shopfloor: add option to exclude package types in product info

### PR DESCRIPTION
**Pull request by JuMiSanAr**
Depends on:
- https://github.com/OCA/product-attribute/pull/1523

Qty picker:

Settings:

ref: cos-4389

### CODE REVIEW

File: shopfloor/models/res_company.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    shopfloor_excluded_product_packaging_ids = fields.Many2many(
```
Comment (simahawk) [inline shopfloor/models/res_company.py]: Why not on the shopfloor app?

File: shopfloor/actions/data.py [hunk 2]
```diff
@@ -309,9 +309,17 @@ def _product_parser(self):
             ("seller_ids:supplier_code", self._product_supplier_code),
         ]
 
-    def _product_packaging(self, rec, field):
+    def _product_packaging(self, rec, field, apply_excluded=True):
```
Comment (simahawk) [inline shopfloor/actions/data.py]: is `apply_excluded` used anywhere?

**Review by simahawk — CHANGES_REQUESTED** [review]

File: shopfloor/actions/data.py (hunk 2 above)
Comment (JuMiSanAr) [reply shopfloor/actions/data.py]: No, I added it here in case we ever need to bypass this new logic.

File: shopfloor/models/res_company.py (hunk 1 above)
Comment (JuMiSanAr) [reply shopfloor/models/res_company.py]: I thought of it, but it seemed to me like a very specific setting compared to the rest of the settings in the shopfloor app (i.e. this doesn't really impact the app as a whole, but a specific part of the shopfloor logic).

I can change it if you prefer.

Comment (simahawk) [reply shopfloor/models/res_company.py]: Thinking again on this: wouldn't make more sense to have a flag on pkg types?
CC @jbaudoux

Comment (mmequignon) [reply shopfloor/models/res_company.py]: I think I would've added a flag at packaging to determine if the package is a vendor package or not.
This way we can filter this depending on the scenario / menu.
T'would probably make it easier to do smth like that `_product_packaging(rec, ..., show_vendor_pkgs=False)`.
show_vendor_pkgs taking the value of `menu.show_vendor_pkgs`.

Comment (jbaudoux) [reply shopfloor/models/res_company.py]: I think it is better to have this concept of external package outside shopfloor as we also have a module to compute qties on a move by package type. We also don't want to include vendor pallet on non incoming moves.
I like @mmequignon proposal. The flag must be at packaging type and the menu could configure if we include them or not. For instance, on the reception, we could be interested to see the quantity of the vendor pallet.

Comment (JuMiSanAr) [reply shopfloor/models/res_company.py]: I've created a new module to enable this option here:
https://github.com/OCA/product-attribute/pull/1523

Comment (JuMiSanAr) [reply shopfloor/models/res_company.py]: I have now updated this PR to use the option in the PR in my last comment above.

Now the vendor packagings will not be displayed by default, and I enabled it in reception as discussed.

Comment (mmequignon) [reply shopfloor/models/res_company.py]: I believe reception is an example, we need to determine where we want to see those packaging.

File: shopfloor/actions/data.py [hunk 3]
```diff
@@ -310,8 +326,16 @@ def _product_parser(self):
         ]
 
     def _product_packaging(self, rec, field):
+        display_vendor_packagings = rec.env.context["display_vendor_packagings"]
+        packagings = rec.packaging_ids
+        if display_vendor_packagings:
+            packagings = packagings.filtered(lambda x: x.qty)
+        else:
+            packagings = packagings.filtered(
+                lambda x: x.qty and not x.packaging_type_id.is_vendor_packaging
```
Comment (jbaudoux) [inline shopfloor/actions/data.py]: This requires a new dependency to https://github.com/OCA/product-attribute/pull/1523 that you forgot to add.

File: shopfloor/actions/data.py [hunk 4]
```diff
@@ -310,8 +326,16 @@ def _product_parser(self):
         ]
 
     def _product_packaging(self, rec, field):
+        display_vendor_packagings = rec.env.context["display_vendor_packagings"]
```
Comment (jbaudoux) [inline shopfloor/actions/data.py]:
```suggestion
        display_vendor_packagings = rec.env.context.get("display_vendor_packagings")
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]
We should not add new dependencies in shopfloor. So those changes should move to a new module shopfloor_vendor_packaging (auto install glue between shopfloor and https://github.com/OCA/product-attribute/pull/1523)

I thought you could transmit the option to the front and let the front manage the exclusion at https://github.com/OCA/wms/blob/14.0/shopfloor_mobile/static/wms/src/wms_utils.js#L328
without a global variable in the context. All those overrides to manage context seem fragile.

**Comment by JuMiSanAr** [comment]
> We should not add new dependencies in shopfloor. So those changes should move to a new module shopfloor_vendor_packaging (auto install glue between shopfloor and [OCA/product-attribute#1523](https://github.com/OCA/product-attribute/pull/1523))
> 
> I thought you could transmit the option to the front and let the front manage the exclusion at https://github.com/OCA/wms/blob/14.0/shopfloor_mobile/static/wms/src/wms_utils.js#L328 without a global variable in the context. All those overrides to manage context seem fragile.

I've moved this to the new `shopfloor_vendor_packaging` module.

Regarding trasmitting the option to the front: as far as I understood it (and as far as I see it), we should not let the frontend decide these things. The qty picker will in the end display whatever the backend sends.

**Review by mmequignon — CHANGES_REQUESTED** [review]
This will be valid only if this module is installed.
I believe this has its place in `shopfloor`, adding a default behavior: do not display vendor packages.
Then, we add this option on the menus we want.

**Comment by JuMiSanAr** [comment]
> This will be valid only if this module is installed. I believe this has its place in `shopfloor`, adding a default behavior: do not display vendor packages. Then, we add this option on the menus we want.

This module will be installed automatically when shopfloor is installed right? At least that's what I understand that happens when the `auto-install` flag is set to True.

**Review by mmequignon — CHANGES_REQUESTED** [review]
Add unit tests please :pray:

**Comment by mmequignon** [comment]
> > This will be valid only if this module is installed. I believe this has its place in `shopfloor`, adding a default behavior: do not display vendor packages. Then, we add this option on the menus we want.
> 
> This module will be installed automatically when shopfloor is installed right? At least that's what I understand that happens when the `auto-install` flag is set to True.

well, yes, you're right, I missed this.

@jbaudoux @simahawk any insight about this?

My thoughts are that shopfloor is meant implement incoming/outgoing/internal flows.
To me it makes sense that this is a core feature.
on the other hand, that would be another dependency.
Thanks!

File: shopfloor_reception/__manifest__.py [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
+    "version": "14.0.2.7.0",
     "development_status": "Beta",
     "category": "Inventory",
     "website": "https://github.com/OCA/wms",
     "author": "Camptocamp, Odoo Community Association (OCA)",
     "maintainers": ["mmequignon", "JuMiSanAr"],
     "license": "AGPL-3",
     "installable": True,
-    "depends": ["shopfloor"],
+    "depends": ["shopfloor", "shopfloor_vendor_packaging"],
```
Comment (JuMiSanAr) [inline shopfloor_reception/__manifest__.py]: Move to its own `shopfloor_reception_vendor_packaging`

File: shopfloor_vendor_packaging/models/shopfloor_menu.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+class ShopfloorMenu(models.Model):
+    _inherit = "shopfloor.menu"
+
+    display_vendor_packagings = fields.Boolean(
+        string="Display vendor packagings",
+        default=False,
+        help=DISPLAY_VENDOR_PACKAGINGS_HELP,
+    )
+    display_vendor_packagings_is_possible = fields.Boolean(
```
Comment (mmequignon) [inline shopfloor_vendor_packaging/models/shopfloor_menu.py]: no plurals in names :pray:

File: shopfloor_vendor_packaging/models/shopfloor_menu.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
+# Copyright 2024 Camptocamp SA (http://www.camptocamp.com)
```
Comment (mmequignon) [inline shopfloor_vendor_packaging/models/shopfloor_menu.py:1]: file not imported. Missing `__init__.py`

File: shopfloor_vendor_packaging/__init__.py [hunk 8]
```diff
@@ -0,0 +1,2 @@
+from . import models
+from . import actions
```
Comment (mmequignon) [inline shopfloor_vendor_packaging/__init__.py:2]: from .
```suggestion
from . import actions
from . import models
```
you would've noticed if there were tests :)

**Review by mmequignon — CHANGES_REQUESTED** [review]

File: shopfloor_reception/migrations/14.0.2.7.0/post-migration.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+from odoo import SUPERUSER_ID, api
+
+_logger = logging.getLogger(__name__)
+
+
+def migrate(cr, version):
+    if not version:
+        return
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    reception_scenario = env["shopfloor.scenario"].search([("name", "=", "Reception")])
```
Comment (simahawk) [inline shopfloor_reception/migrations/14.0.2.7.0/post-migration.py]:
```suggestion
    reception_scenario = env["shopfloor.scenario"].search([("key", "=", "reception")])
```

File: shopfloor_reception/migrations/14.0.2.7.0/post-migration.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    options = scenario.options
+    options["display_vendor_packagings"] = True
+    options_edit = json.dumps(options or {}, indent=4, sort_keys=True)
+    scenario.write({"options_edit": options_edit})
+    _logger.info("Option display_vendor_packagings added to scenario Zone Picking")
+
+
+def _enable_option_in_menus(menus):
+    for menu in menus:
+        menu.display_vendor_packagings = True
```
Comment (simahawk) [inline shopfloor_reception/migrations/14.0.2.7.0/post-migration.py]:
```suggestion
        menus.display_vendor_packagings = True
```

File: shopfloor_reception/services/reception.py [hunk 11]
```diff
@@ -710,9 +710,13 @@ def _data_for_stock_pickings(self, pickings, with_lines=False):
         ]
 
     def _data_for_move_lines(self, lines, **kw):
+        if self.work.menu.display_vendor_packagings:
+            kw["display_vendor_packagings"] = True
```
Comment (simahawk) [inline shopfloor_reception/services/reception.py]:
```suggestion
            kw["display_vendor_packagings"] = self.work.menu.display_vendor_packagings
```

File: shopfloor_vendor_packaging/__manifest__.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    "category": "Inventory",
+    "website": "https://github.com/OCA/wms",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "application": False,
+    "data": [
+        "views/shopfloor_menu.xml",
+    ],
+    "depends": ["shopfloor", "product_packaging_type_vendor"],
+    "auto-install": True,
```
Comment (simahawk) [inline shopfloor_vendor_packaging/__manifest__.py:18]:
```suggestion
    "auto_install": True,
```
it won't work otherwise

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by simahawk** [comment]
> > > This will be valid only if this module is installed. I believe this has its place in `shopfloor`, adding a default behavior: do not display vendor packages. Then, we add this option on the menus we want.
> > 
> > 
> > This module will be installed automatically when shopfloor is installed right? At least that's what I understand that happens when the `auto-install` flag is set to True.
> 
> well, yes, you're right, I missed this.
> 
> @jbaudoux @simahawk any insight about this?
> 
> My thoughts are that shopfloor is meant implement incoming/outgoing/internal flows. To me it makes sense that this is a core feature. on the other hand, that would be another dependency. Thanks!

I don't think is a core feature, but I don't know if this is should be restricted to reception only.
My take: if we are unsure today, let's keep it like this and then we'll see.
@jbaudoux WDYT?

**Review by jbaudoux — APPROVED** [review]

**Review by jbaudoux — COMMENTED** [review]
I approve on the principle but you'll have to squash some commits and split with one commit per module

**Review by simahawk — APPROVED** [review]

**Comment by jbaudoux** [comment]
Can you rebase ?

**Comment by JuMiSanAr** [comment]
> Can you rebase ?

Done

**Pull request merged**